core/tx_pool.go

```diff
@@ -1266,6 +1265,14 @@ func (pool *TxPool) reset(oldHead, newHead *types.Header) {
 	pool.pendingNonces = newTxNoncer(statedb)
 	pool.currentMaxGas = CalcGasLimit(pool.chain.CurrentBlock(), statedb)
 
+	// we don't replace currentCtx, since it's shared with all txLists
```

Comment:
If `currentCtx` is shared (which means `BlockContext` and `CurrencyManager` probably are as well, do they handle concurrency properly? If they do, documenting that is important)

Reply:
This data sharing makes me a little nervous. In general I think its a bad idea to mutate a shared object. (In Rust, this would not be allowed at all.) It would be better is the tx lists did not hold a pointer to the context object and instead had it provided in the function arguments where needed.

Reply:
yeah, agree. txpool on a few places (goroutines) requires this, also txlist and txpricedlist... so i'll see if instead of injecting the context i'll inject a context provider that has a lock

Reply:
Check https://github.com/celo-org/celo-blockchain/pull/1499/commits/e7aa53d5ed903e53e47c39ca76b050f8e10508cc where i'm using `atomic.Value` to solve the concurrency problem.